Copy DQM ROOT file to EOS and register entry in DQMUpload executor

khurtado opened the pull request:
Fixes #10287

#### Status
New changes being tested

#### Description
This implements the new DQM Gui method described in #10287

- Copy DQM root file to EOS
- Register entry to new DQMGUI: Note this does not upload any files. It registers one root file per call, right after the file has been uploaded to EOS.  
- Delete files uploaded to EOS if registration fails
- Use value of forceRunNumber for multiRun, to replicate current behavior by old visDQMGUI

#### Is it backward compatible (if not, which system it affects?)
YES

Commit:
Support new DQMGUI.
- Copy DQM root files to EOS
- Register files to new DQMGUI via http post method
- If registration fails, delete EOS files upload

khurtado commented:
@amaltaro I think this *almost is ready for review. What is pending before the code review is getting a new permanent EOS area, managed by used, rather than the unmerged area.

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -50,6 +56,14 @@ class DQMUpload(Executor):
+        self.retryCount = 3
+        self.registerLFNBase = '/store/unmerged/DQMGUI'
+        self.registerEOSPrefix = '/eos/cms'
+        self.registerURL = 'https://cmsweb-testbed.cern.ch/dqm/offline-test-new/api/v1/register'
```
Can we derive this URL from the url used for uploading the root file to the DQM Gui?

khurtado replied:
Do you meant `step.upload.URL`? Aren't they independent?

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+        eosStageOutParams['option'] = overrides.get('option', "--wma-disablewriterecovery")
+        eosStageOutParams['phedex-node'] = overrides.get('eos-phedex-node', "T2_CH_CERN")
+        eosStageOutParams['lfn-prefix'] = overrides.get('eos-lfn-prefix',
+                                                         "root://eoscms.cern.ch/%s" % self.registerEOSPrefix)
```
It seems not to be indented with the line above(?)

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+        eosFileInfo = {'LFN': self.getAnalysisFileLFN(analysisFile),
+                       'PFN': os.path.join(stepLocation, 
+                                           os.path.basename(analysisFile.fileName)),
+                       'PNN': None,
```
Should we fill this PNN out? Or it gets filled during the underlying stage out modules?

khurtado replied:
It's taken care by the `StageOutMgr`. Maybe I can add a comment with that?

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+
+    def _register(self, registerURL, args):
+        """
+        POST request to register URL
```
Please update this docstring as well.

amaltaro replied:
BTW, I think we should merge this code with what we have under the `upload` method to avoid duplication. Maybe create a new method named like `_setupHttp`.

khurtado replied:
I thought about that at first but even though both are POST requests, the `upload` method actually uploads file data content, while this doesn't, so I made them separated to distinguish the original purpose of each. What do you think?

amaltaro replied:
Maybe we could have something like `def _getHttpsOpener` with the initial common code? It it becomes too ugly, then we keep these into 2 different methods.

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+            [{"dataset": "/a/b/c", "run": "123456", "lumi": "0", "file": "/eos/cms/store/group/comm_dqm/DQMGUI_data/location/file.root", "fileformat": 1}]
+        Set lumis to 0, as this is reproducing current per Run based root files
+        - https://github.com/dmwm/WMCore/issues/10287#issuecomment-1052558061
+        DQM root files produced by Harvesting prosessing are type 1 (plain ROOT)
```
typo in `processing`

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+        - https://github.com/cms-DQM/dqmgui#file-formats
+        """
+        # Get task description
+        job = Bootstrap.loadJobDefinition()
```
Kenyi, could you please check whether the job description is already available in this object? From:
https://github.com/dmwm/WMCore/blob/ee8c9fe76301997917e938b998393760893de7cf/src/python/WMCore/WMSpec/Steps/Executor.py#L93

it might be that we already have it under `self.job`

khurtado replied:
Good catch! I will check

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+                raise WMExecutionFailure(70320, "DQMUploadFailure", msg)    
+
+        args = {}
+        args['file'] = self.registerEOSPrefix + \
```
It might be safer to use `os.path.join` to avoid issues with double or lack of slashes. The line looks short enough as well, maybe put it all in the same line.

khurtado replied:
Same as one comment above, I would need to remove the starting `/` from the second argument in order to use `os.path.join`, otherwise we get:

```
>>> os.path.join("/eos/cms","/store/user/something")
'/store/user/something'
```

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+
+        try:
+            logException = True
+            for registerURL in self.registerURL.split(';'):
```
We should contact the DQM Gui and see whether the file registration should be done against a single server (for each workflow). If so, I'm very much in favor of removing this. Actually, it was added to the DQMUpload step many many years ago, I am not even sure whether it's still required and useful to anyone.

If we can remove the multi url manipulation, then we can likely try to remove this `logException` flag as well.

khurtado replied:
Good point! I just left a comment in the issue asking the DQM folks.

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+                else:
+                    msg = 'HTTP POST to register url finished succesfully with response:\n' + msg
+                    logging.info(msg)
+        except Exception as ex:
```
Given that we have a retry logic for the root file stage out, I think we should recycle the same logic (or something similar) to this POST requests. If we get this far, then we should try as hard as we can not to fail this job.

khurtado replied:
That's fair. I will implement some retrial feature for this method

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+
+        logging.info("Deleting LFN: %s", eosFileLFN)
+        eosFileInfo = {'LFN': eosFileLFN,
+                           'PFN': None, # PFNs are assigned in the Delete Manager
```
Indentation needs to be fixed.

khurtado replied:
Ah, I wonder how that happened. Thanks, will do

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+        except Alarm:
+            msg = "Indefinite hang while deleting file from EOS"
+            logging.error(msg)
+            raise WMExecutionFailure(70321, "DQMUploadStageOutTimeout", msg)
```
to avoid a long and unreadable traceback, I'd suggest to complete this raise line with `from None`. Same for the block below.

amaltaro reviewed (changes requested):
Kenyi, I left some comments along the code. And we might want to contact the DQM team as well to see what we have to support in terms of urls (do we need to support multi urls?)

khurtado commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+        """
+        # If lfn start with '/', make it relative to it
+        lfnFile = os.path.relpath(analysisFile.lfn, start='/')
+        lfn = os.path.join(self.registerLFNBase, lfnFile)
```
I tried that first, but was surprised with the `self.registerLFNBase` being ignored if `analysisFile.lfn` started with `/`:

```
>>> os.path.join("/root", "/hello")
'/hello'
```

which is why I had to make sure the `analysisFile.lfn` has a relative form

amaltaro replied:
and this is a surprise to me as well!

khurtado commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+        """
+        Copy DQM Root files to EOS and register files to new DQMGUI
+        """
+        eosFileLFN = self._uploadToEOS(analysisFile, stepLocation)
```
1. But `_uploadToEOS` finishes with a try/except block already, so it throws an exception if anything fails here:

https://github.com/dmwm/WMCore/blob/ee8c9fe76301997917e938b998393760893de7cf/src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py#L356-L365
2.  Sounds good to me, I will switch the position of the arguments

Commit:
Address PR comments
- Fix indentation
- Assume single register URL
- Retry registration
- Use self.job directly

khurtado commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +309,229 @@ def upload(self, url, args, filename):
+
+    def uploadToEOSAndRegister(self, stepName, stepLocation, analysisFile):
+        """
+        Copy DQM Root files to EOS and register files to new DQMGUI
```
Done

khurtado commented:
@amaltaro I think I addressed most of the comments and left comments on the ones I didn't

amaltaro reviewed (commented):
Kenyi, sorry for the delay reviewing this. Next time, please refresh the review request through GH such that we can properly filter it.

While reviewing it, it just occurred to me that T0 also uses this code. So we need to verify whether T0 would need to use this same mechanism; or whether we need to work on workflow configuration that could enable/disable this new DQM feature. Can you please follow this up with the DQM team?

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +311,253 @@ def upload(self, url, args, filename):
+                        logging.exception(msg)
+                    else:
+                        logging.error(msg)
+                    raise WMExecutionFailure(70319, "DQMUploadFailure", msg)
```
Please add this new error code and message to the WMExceptions.py module as well.

amaltaro commented on src/python/WMCore/WMSpec/Steps/Executors/DQMUpload.py:
```diff
@@ -293,3 +311,253 @@ def upload(self, url, args, filename):
+                    logging.info(msg)
+                    break
+            except Exception as ex:
+                if numRetry == self.retryCount:
```
Maybe we should have something like:
```
if numRetry < self.retryCount:
    print "registration failed blah blah"
    sleep
else:
   logger.exception(blah)
```
what do you think?

amaltaro reviewed (changes requested):
@khurtado Kenyi, here goes a fresh review of the current changes. In addition to the comments made along the code, I'd like to mention:
1. Please update the initial PR description.
2. I think it would be extremely useful to have a short summary of what DQMUpload module does at the top of the module, including all the steps it takes (since it no longer does a simple upload of the root file).
3. We also need to clarify whether T0 workflows will have to follow the same new design (have you heard anything from the DQM team?)
4. Did we agree on a final map of urls? If so, it needs to be implemented as well.

Commit:
- Use new EOS area
- Use upload.URL -> registerURL mapping
- Address PR comments

khurtado commented:
Hi @amaltaro , to answer your questions:

1. Do you mean the first commit, or this PR GH description?
2. Agreed, Done!
3. Yes, it's the same machinery for T0, answer 1 in: https://github.com/dmwm/WMCore/issues/10287#issuecomment-1086730224
4. Done

amaltaro commented:
> 1. Do you mean the first commit, or this PR GH description?

The initial PR description at the very top, here: https://github.com/dmwm/WMCore/pull/11015#issue-1154640531
However, having a second look at it, it looks good to me. Perhaps can you expand on the granularity used for the file registration in the new DQM Gui? Is it one HTTP call for each run? For each root file?

amaltaro commented on src/python/WMCore/WMExceptions.py:
```diff
@@ -184,7 +184,12 @@
                       60451: "Output file lacked adler32 checksum (WMAgent).",  # (WMA)
+                      70317: "Indefinite hangout while staging out DQM root file to EOS",  # (WMA)
                       70318: "Failure in DQM upload.",
+                      70319: "HTTP POST to register url failed.",  # (WMA)
```
I'd suggest to rephrase it to ""HTTP POST to DQM register url failed.""

amaltaro reviewed (approved):
Kenyi, other than the minor exception suggestion, it looks good to me.

khurtado commented:
> Kenyi, other than the minor exception suggestion, it looks good to me.
@amaltaro Thanks! I just made that last change. 
Note we still need to have a successful test with the new changes. It is currently failing while registering the file, but it looks like it's due to cmsweb related issue with the new mapping rather than this code. See my last comment below:

https://github.com/dmwm/WMCore/issues/10287#issuecomment-1119818542
